### PR TITLE
Simplify `add_distributions` in `_SearchSpaceGroup`

### PR DESCRIPTION
**Pull request by not522**
## Motivation
This is a follow-up PR of #2526. It replaces the recursive implementation of `_add_distributions` with a non-recursive one.

## Description of the changes
- Implement non-recursive `add_distributions`
- Remove the implementation detail from the document.

### CODE REVIEW

**Comment by HideakiImamura** [comment]
@hvy Could you review this PR if you have time?

File: optuna/samplers/_tpe/sampler.py [hunk 1]
```diff
@@ -158,33 +158,6 @@ def objective(trial):
...
-            - If an element ``S`` of a group contains ``T``, remove ``S`` from the group and add
-              ``T`` and ``S-T`` to the group.
-            - If the intersection of an element ``S`` of the group and ``T`` is not empty, remove
-              ``S`` from the group and add ``S∩T``, ``S-T``, and ``T-S`` to the group.
-              We recursively add ``T-S`` to the group because the intersection of ``T-S`` and
-              some other elements of the group may not be empty.
-
-            The group of the search space recursively constructed based on the above rules are
-            disjoint and the union is the entire search space. We perform sampling from the joint
-            distribution for each element of this decomposed group of the search space.
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/sampler.py:186]: How about leaving a note as follows?

```
The group of the search space is disjoint and the union is the entire search space. 
We perform sampling from the joint distribution for each element of this decomposed 
group of the search space.
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the simplification! Almost LGTM. I have a comment around the deleted docs. Please take a look.

**Review by hvy — APPROVED** [review]
Nice, LGTM!

File: optuna/samplers/_tpe/sampler.py (hunk 1 above)
Comment (HideakiImamura) [reply optuna/samplers/_tpe/sampler.py:186]: Ops. Thanks for the clarification. You are right. Let me resolve this conversation.

**Pull request merged**